proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,432 @@
+In particular, this can affect algorithms involving shadow maps and any case
+ where standard samples have comparison sample counterparts that need to match.
+In addition, this represents a discrepancy with comparable platforms where this
+ deficiency makes porting, emulating, or developing for multiple platforms
+ difficult or impossible.
+
+## Proposed solution
+
+To add full comparison sampling support, HLSL will introduce an optional feature
+ that adds these new comparison sampling builtin methods to all texture objects
+ that currently support the non-comparison methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias)`
+  performs a comparison sample using the implicitly-determined LOD with the
+  explicitly-provided _Bias_ value added to that result.
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY)`
+   samples using the LOD calculated using the explicitly-provided
+   _DDX_ and _DDY_ gradients.
+
+Only the simplest overloads of the new and existing builtin methods are
+ described in detail here for the sake of brevity,
+Though not included above, all relevant overloads of `SampleBias` and
+ `SampleGrad` will be reflected in their new `SampleCmpBias` and
+ `SampleCmpGrad` counterparts.
+The parameters not explicitly described here will function exactly as they do
+ when used by the non-comparison builtin methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset, float Clamp)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset, float Clamp, uint Status)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset, float Clamp)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset, float Clamp, uint Status)`
+
+To add LOD calculation for comparison samplers, HLSL will introduce and require
+ these new comparison sampler method overloads to all texture objects that
+ currently support the non-comparison method overloads:
+
+* `CalculateLevelOfDetailUnclamped(SamplerComparisonState S, float<num> Location)`
+  returns the raw LOD value determined implicitly using gradients representing
+  how much the _Location_ changes across neighboring pixels regardless of
+  whether it falls outside the valid range.
+* `CalculateLevelOfDetail(SamplerComparisonState S, float<num> Location)`
+  returns the implicitly-determined LOD value clamped to the valid range.
+
+## Detailed Design
+
+### HLSL additions
+
+These new HLSL texture methods will be added:
+
+```c++
+<Format> <TexObject>::SampleCmpGrad(
```

Comment:
Can we write these as more complete pseudo-c++? Something like:

```c++
template<typename ElementType, int ElementSize>
vector<ElementType, ElementSize> TextureBase::SampleCmpGrad(
    SamplerComparisonState S,
    vector<float,ElementSize> Location,
    ...);
```

In the Work Graphs spec I tried to be explicit about adding full class hierarchies in pseudo code to make it easier to see exactly what interfaces are being added on which types, what the return types are, and how they relate to the templated parameters.